Update IC_BrivGemFarm_Functions.ahk

Klaws-- opened the pull request:
*Experimental* change to improve reliability/recovery if memory reads go wrong. Previously, if the SB stack count cannot be read correctly, offline stacking will enter an infinite loop. There was a limit of 19 retries maximum, but it didn't work. This change reduces the number of allowed retries to 2 - which may be okay for a lot of users, and possibly totally NOT okay for others.

Consider this commit more as "food for thought", not as a proposed change. Yes, you are supposed to reject the pull request.

antilectual commented:
Thanks for the find. There is definitely a bit of a bug here. The script does do 10 retries before continuing with the script, but continuing with the script in turn decides stacks are low and it's time to stack. It does 10 more tries and that's where the infinite loop comes in.

Leaving the retries at 10 for now. Whether it's 10 or 2 retries, a run is effectively ruined by being unable to stack. 10 should give players more time to see something is wrong before the script restarts their run with no stacks for jumping. 

Adding the retry number (as well as a short warning) to the restart text seems like a prudent thing to do. It's common for new players to have settings that cause multiple restacks and having an indication that it is not normal behavior early should help get them pointed in the right direction.

Klaws-- commented:
Thanks for your review!

I updated the code by adding a few comments so I can remember what my changes do ;) and also changed the max # of retries to 10 (it sure made testing a bit less boring by having it at 2, but, as we both mentioned, that's probably not what we want in the "prod version").

Sorry for delaying my edit; I just started on a new job and I also suffer from the nasty and uncontrollabe urge to test my changes before I commit them - and yes, I did test again, even though "what could possibly go wrong when just adding comments and changing a number?" ;-)

Note that this is the first time I amend to a pull request og GitHub, hope it worked. That's the commit whoch should jave arrived here: https://github.com/Klaws--/Idle-Champions/commit/4e62b665357e737013877b6b95541058a6608cc4

antilectual commented:
No worries about the delay. 
I've been working on this issue but instead of disabling stacking entirely, if it fails to stack a certain number of times it'll automatically increment the stack zone until it gets to one where it gets stacks (or resets, whichever comes first).
If it fails stacking at the zone set in settings for multiple runs, it switches from settings to using whatever zone it was successful on.

Hopefully this will help with people losing their gem farm to seasonal buffs or with new players setting their stack zone a touch too low. 

At least that's how I envision it working, but confirming it with tests has been taking a very long time since it requires a lot of failing and restarting the game. At the very least, it tries to push the run forward instead of infinite stacking which is an improvement. The changes should hit the GameObjectRefactor branch "soon". 

Also, good luck with your new job!

Klaws-- commented:
Doing dome more testing...seems that while my "fix" appears to work in some cases, it fails in others. Yep, there are about 8.12e127 different ways steelbone stacking can fail, and my code fails to catch 'em all. It also doesn't help that the game received updates every 5 minutes while I am testing :/

The adaptive approach you describe is a much better solution. Since my experience with ICScriptHub comes from just a cursoray glance at the source code, that's nothing I could tackle with any confidence in success.

Talking of game game updates...I had to hotfix some code in IC_IdleGameManager64_Import.ahk to keep going:

```
;this.game.gameInstances.Controller.userData.StatHandler.BrivSteelbonesStacks := New GameObjectStructure(this.game.gameInstances.Controller.userData.StatHandler,"Int", [0x300])
;this.game.gameInstances.Controller.userData.StatHandler.BrivSprintStacks := New GameObjectStructure(this.game.gameInstances.Controller.userData.StatHandler,"Int", [0x304])
this.game.gameInstances.Controller.userData.StatHandler.BrivSteelbonesStacks := New GameObjectStructure(this.game.gameInstances.Controller.userData.StatHandler,"Int", [0x310])
this.game.gameInstances.Controller.userData.StatHandler.BrivSprintStacks := New GameObjectStructure(this.game.gameInstances.Controller.userData.StatHandler,"Int", [0x314])
```

Epic Game Store delivered version 507 to me now. Other offsets are probably f*ed up as well, haven't checked though. Since CNE mentioned that different platforms may have different versions; I wonder how such changes could be implemented without breaking platforms which lag behind...I mean, sure, defining the imports for each game version and automatically picking the right ones...ah well, sounds like no fun at all.

Thanks for putting so much effort into this tool! Very much appreciated!

antilectual commented:
Do you know which cases are causing you issues? I'd like to make sure my changes have covered those cases as well.

Updating imports is something I no longer do on each release or even on any sort of fixed schedule. That is the purpose behind [Script Hub Importer](https://github.com/antilectual/ScriptHub-AutomaticOffsets). 
I know not everyone is capable or willing to use it, but the official [Discord](https://discord.com/invite/idlechampions/?utm_source=landing%20page&utm_medium=General&utm_campaign=general&utm_content=Discord) has a pinned thread in the ``#scripting`` channel dedicated to updated Imports that users share. 

So far there has only been one case of platforms not using the same Imports for the same game version when EGS v505 had to update Imports when Steam v505 was using the same Imports as Steam v504.

Klaws-- commented:
Not sure how it's triggered...but I've got a few observations.

1. this-StackFarm() is called at more than one place.
2. reatryAttempt is a local variable in StackRestart().

So I conclude that, if StartRestart() get called multiple times per run, each time successful (that is, reatyAttempts always stays below 10, so this.noStacking (the variable I added) is never set), it can still be called be multiple times per run, each time displaying "StackRestart #1" (never getting to StarckRestart #10).

Again, not sure how this happens. But I guess retryAttempts should instead be an instance variable like this.stackingRetryAttempts, which gets reset when the script detects area/zone 1. Naturally, this will need to get checked at every decision point where stacking might happen. It will also take the place of this.noSTacking that way, making this.noStacking obsolete.

Regarding the Dicord channel...was not aware of that. I guess I'm getting too old for this "internet" thing. Oh well, or maybe I'm just better at reading source code written in an unknown programming language than reading documentation...or maybe I just **think** that I'm better at that. ;)
